core/src/main/scala/org/apache/spark/SparkError.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import java.net.URL
+
+import scala.collection.immutable.SortedMap
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.core.`type`.TypeReference
+import com.fasterxml.jackson.databind.SerializationFeature
+import com.fasterxml.jackson.databind.json.JsonMapper
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+
+import org.apache.spark.util.Utils
+
+/**
+ * Information associated with an error class.
+ *
+ * @param sqlState SQLSTATE associated with this class.
+ * @param messageFormatLines C-style message format compatible with printf.
+ *                           The error message is constructed by concatenating the lines with
+ *                           linebreaks.
+ */
+case class ErrorInfo(sqlState: Option[String], messageFormatLines: Seq[String]) {
```

Comment:
This is defined in core. But sql state is a SQL concept?

Reply:
I agree; SQLSTATE is a SQL concept. Off the top of my head, I'm not sure how to have this only be set in the `sql` component without increasing the complexity of the implementation:

- Create a JSON file that maps error classes to SQLSTATE in SQL, which may increase maintenance burden and decrease developer usability
- Create a `SqlErrorInfo(sqlState, messageFormatLines)` class that extends `ErrorInfo(messageFormatLines)` in the `sql` component, and parse the JSON file again to get the `sqlState` field

What do you think?

Reply:
In addition to the implementation difficulty, I think it also makes sense to have `sqlState` be accessible from any type of Spark exception (regardless of the base component), given that we currently throw `SparkException`s (from the `core` component) during query execution. We could introduce a `SparkSqlError` type, but we would have to do a major refactor of the existing exception types as well in that case.

Reply:
It's better to have a unified representation for all the errors. "error class" and "message" are required fields, and it's ok to have more optional fields for other purposes, like "sqlState" for JDBC compatibility.

Reply:
Yup, maybe we can consider giving a general name for `sqlState` (e.g. errorState, errorStateNumber, or etc).

Reply:
Postgres calls their SQLSTATEs [error codes](https://www.postgresql.org/docs/13/errcodes-appendix.html), but I'm not sure if it'd be safe for us to do the same thing, especially given that we could add Spark-specific error codes in the future that span across non-SQL functionality.

Reply:
I prefer `sqlState`, to clearly indicate what it is, since SQL state is a standard.